Cubical/Data/Rationals/MoreRationals/QuoQ/Properties.agda

```diff
@@ -0,0 +1,248 @@
+{-# OPTIONS --safe #-}
+module Cubical.Data.Rationals.MoreRationals.QuoQ.Properties where
+
+open import Cubical.Foundations.Prelude
+open import Cubical.Foundations.Isomorphism
+open import Cubical.Foundations.Equiv
+open import Cubical.Foundations.GroupoidLaws hiding (_⁻¹)
+open import Cubical.Foundations.Univalence
+
+open import Cubical.Data.Int.MoreInts.QuoInt as ℤ using (ℤ; Sign; signed; pos; neg; posneg; sign)
+open import Cubical.HITs.SetQuotients as SetQuotient using () renaming (_/_ to _//_)
+
+open import Cubical.Data.Nat as ℕ using (ℕ; zero; suc)
+open import Cubical.Data.NatPlusOne
+open import Cubical.Data.Sigma
+
+open import Cubical.Data.Sum
+open import Cubical.Relation.Nullary
+
+open import Cubical.Data.Rationals.MoreRationals.QuoQ.Base
+
+ℚ-cancelˡ : ∀ {a b} (c : ℕ₊₁) → [ ℕ₊₁→ℤ c ℤ.· a / c ·₊₁ b ] ≡ [ a / b ]
+ℚ-cancelˡ {a} {b} c = eq/ _ _
+  (cong (ℤ._· ℕ₊₁→ℤ b) (ℤ.·-comm (ℕ₊₁→ℤ c) a) ∙ sym (ℤ.·-assoc a (ℕ₊₁→ℤ c) (ℕ₊₁→ℤ b)))
+
+ℚ-cancelʳ : ∀ {a b} (c : ℕ₊₁) → [ a ℤ.· ℕ₊₁→ℤ c / b ·₊₁ c ] ≡ [ a / b ]
+ℚ-cancelʳ {a} {b} c = eq/ _ _
+  (sym (ℤ.·-assoc a (ℕ₊₁→ℤ c) (ℕ₊₁→ℤ b)) ∙ cong (a ℤ.·_) (ℤ.·-comm (ℕ₊₁→ℤ c) (ℕ₊₁→ℤ b)))
+
+-- useful functions for defining operations on ℚ
+
+onCommonDenom :
+  (g : ℤ × ℕ₊₁ → ℤ × ℕ₊₁ → ℤ)
+  (g-eql : ∀ ((a , b) (c , d) (e , f) : ℤ × ℕ₊₁) (p : a ℤ.· ℕ₊₁→ℤ d ≡ c ℤ.· ℕ₊₁→ℤ b)
+           → ℕ₊₁→ℤ d ℤ.· (g (a , b) (e , f)) ≡ ℕ₊₁→ℤ b ℤ.· (g (c , d) (e , f)))
+  (g-eqr : ∀ ((a , b) (c , d) (e , f) : ℤ × ℕ₊₁) (p : c ℤ.· ℕ₊₁→ℤ f ≡ e ℤ.· ℕ₊₁→ℤ d)
+           → (g (a , b) (c , d)) ℤ.· ℕ₊₁→ℤ f ≡ (g (a , b) (e , f)) ℤ.· ℕ₊₁→ℤ d)
+  → ℚ → ℚ → ℚ
+onCommonDenom g g-eql g-eqr = SetQuotient.rec2 isSetℚ
```

Comment:
Wasn't there an issue with SetQuotient.rec2 being defined by rec instead of double pattern-matching? Maybe this get easier with these definition if that is fixed?

Reply:
I can see about fixing rec2 to see if things compute a bit easier, and if not I can define this with pattern-matching as well.

Reply:
Threw in the change to rec2.